broadinstitute/wfl
# GH-1328 Migrate to new Rawls Snapshot V2 endpoints

### Purpose


- https://broadinstitute.atlassian.net/browse/GH-1328

Switched to Rawls' new [Snapshot V2](https://rawls.dsde-prod.broadinstitute.org/#/snapshots_v2) endpoints, accommodating changes to response format.

Also got to take advantage of the option to [specify a snapshot ID when fetching a list of snapshots](https://rawls.dsde-prod.broadinstitute.org/#/snapshots_v2/enumerateDataRepoSnapshot_v2).  Our only use for listing snapshots previously was as a fallback in the Terra Executor update loop, to fetch a snapshot reference that had already been created.  We would paginate through all of a workspace's snapshot references to find one matching the snapshot ID.

### Changes


Rawls Snapshot V2 responses represent individual snapshot references as follows:
```
{
      "attributes": {
        "instanceName": "terra",
        "snapshot": "67a2bfd7-88e4-4adf-9e41-9b0d04fb32ea"
      },
      "metadata": {
        "cloningInstructions": "COPY_NOTHING",
        "description": "",
        "name": "CDC_COVID19_Surveillance0bb700a0e161486aaf83ca6abbb72ff8",
        "resourceId": "f6a4c088-2143-4bc8-8e93-81e046d8eb55",
        "resourceType": "DATA_REPO_SNAPSHOT",
        "stewardshipType": "REFERENCED",
        "workspaceId": "6f417963-402d-4883-937c-49aebd114859"
      }
    }
```

When listed, the snapshot references have the same format as above and are found at `gcpDataRepoSnapshots` key:
```
{
  "gcpDataRepoSnapshots": [
    <snapshot-reference-1>,
    <snapshot-reference-2>,
    …
  ]
}
```

### Review Instructions


All automated tests passed.

Manual testing:

Triggered a retry on a local COVID workload.
```
curl -X POST "http://localhost:3000/api/v1/workload/e66c86b2-120d-4f7f-9c3a-b9eaadeb1919/retry" \
        -H 'Authorization: Bearer '$(gcloud auth print-access-token) \
        -H "Content-Type: application/json" \
        -d '{ "status": "Failed" }'
```

Observed that new workflow was successfully linked to the old, with no issues obtaining the snapshot reference.
<img width="1561" alt="Screen Shot 2021-08-13 at 11 18 57 AM" src="https://user-images.githubusercontent.com/79769153/129389343-990212d4-a740-42cf-b341-1073f0d5e2ac.png">

Created and started a local COVID workload.
```
curl -X POST "http://localhost:3000/api/v1/create" \
    -H "Authorization: Bearer "$(gcloud auth print-access-token) \
    -H "Accept: application/json" \
    -H "Content-Type: application/json" \
    -d @/Users/okotsopo/Library/Application\ Support/JetBrains/IdeaIC2020.3/scratches/GH-1340/snapshot-workload-request.json
{
  …
  "uuid" : "7c312c96-bcac-4880-b3ea-90579f27790f",
  "executor" : {
      "workspace" : "wfl-dev/CDC_Viral_Sequencing_okotsopo_20210707",
      "methodConfiguration" : "wfl-dev/sarscov2_illumina_full",
      "methodConfigurationVersion" : 44,
      "fromSource" : "importSnapshot",
      "name" : "Terra"
    },
  …
}

curl -X POST "http://localhost:3000/api/v1/start" \
    -H "Authorization: Bearer "$(gcloud auth print-access-token) \
    -H "Content-Type: application/json"
    -d $'{ "uuid": "7c312c96-bcac-4880-b3ea-90579f27790f" }'
```

Observed that the expected single workflow was triggered, with no issues creating the snapshot reference.
<img width="1559" alt="Screen Shot 2021-08-13 at 11 41 11 AM" src="https://user-images.githubusercontent.com/79769153/129391941-5bee8132-26e6-4e7f-855f-dc8ece307ed1.png">
